query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,529 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::Schema;
+
+use mentat_query::{
+    NotJoin, 
+    UnifyVars, 
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+    ErrorKind,
+};
+
+use types::{
+    ColumnConstraint, 
+    ComputedTable,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let unified = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+
+        let mut template = self.use_as_template(&unified);
+        
+        for v in unified.iter() {
+            if self.value_bindings.contains_key(&v) {
+                let val = self.value_bindings.get(&v).unwrap().clone();
+                template.value_bindings.insert(v.clone(), val);
+            } else if !self.column_bindings.contains_key(&v) {
+                bail!(ErrorKind::UnboundVariable(v.name()));
+            }
+        }
+
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+        }
+
+        if template.is_known_empty() {
+            return Ok(());
+        }
+
+        for v in unified.iter() {
+           if self.column_bindings.contains_key(&v) {
```

Comment:
And note that if you do this _before_ you apply clauses — which you should — then you can simply unilaterally insert into `column_bindings`, because you know it's a blank template.